internal/server/boltdbstate/job.go

```diff
@@ -1709,6 +1724,135 @@ func (s *State) taskComplete(jobId string) error {
 	return nil
 }
 
+// pipelineComplete will look up the referenced job to see if it has a PipelineStep ref
+// associated with it, and if so, progress the pipelineRun state machine.
+func (s *State) pipelineComplete(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab the pipeline run
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
```

Comment:
It looks like we do this pattern (grab a job by id, then get the associated pipeline run) for Complete, Ack, and Cancel. I think it makes sense to create a specific boltdb transaction func that does this for us. Perhaps inside `boltdbstate/pipeline_run.go`? Since these funcs might be called often it makes sense I think to do these queries together. Maybe something like `GetPipelineRunByJobId`?

Reply:
Ok, good idea. At first I thought you meant we should look up from the database by JobId, and didn't want to do that since that'd be expensive... and we would need to look through the jobs field on each pipeline and check if the job ID is in it, and I don't think you can index by an array that constantly changes. Buuuut if you meant just a function that wraps the two, that works. I feel awkward putting it in boltdbstate though, since it's more of a mediator function... but also can't think of a better place to put it.